core/src/main/java/com/linecorp/armeria/server/Routers.java

```diff
@@ -71,13 +77,41 @@
                             RejectedRouteHandler.class.getSimpleName(), e);
             }
         };
+
+        final Map<Route, Route> newFallbackRoutes = new HashMap<>();
+        final BiFunction<Route, ServiceConfig, ServiceConfig> fallbackValueConfigurator =
+                (originalRoute, fallbackServiceConfig) -> {
+                    final Route fallbackRoute = fallbackServiceConfig.route();
+
+                    if (originalRoute.complexity() == fallbackRoute.complexity() &&
+                        originalRoute.methods().containsAll(fallbackRoute.methods())) {
+                        return fallbackServiceConfig;
+                    }
+
+                    checkState(fallbackRoute.equals(Route.ofCatchAll()),
+                               "Fallback service must catch all requests.");
+                    final Route newFallbackRoute =
+                            originalRoute.toBuilder()
+                                         .pathMapping(CatchAllPathMapping.INSTANCE)
+                                         // Do not propagate an exception raised while resolving a route
+                                         // to fallback services.
+                                         .canSetDeferredException(false)
+                                         .build();
+                    final Route cachedRoute = newFallbackRoutes.get(newFallbackRoute);
+                    if (cachedRoute != null) {
+                        return fallbackServiceConfig.withRoute(cachedRoute);
+                    }
+                    newFallbackRoutes.put(newFallbackRoute, newFallbackRoute);
+                    return fallbackServiceConfig.withRoute(newFallbackRoute);
```

Comment:
Maybe this is wrong.
- This function maybe be called in multiple threads.
- On the second thought(😅), is it better to cache `fallbackServiceConfig`s?
```java
final Map<Route, ServiceConfig> newFallbackServiceConfigs = new ConcurrentHashMap<>();
final BiFunction<Route, ServiceConfig, ServiceConfig> fallbackValueConfigurator =
        (originalRoute, fallbackServiceConfig) -> {

final ServiceConfig cachedFallbackConfig = newFallbackServiceConfigs.get(originalRoute);
if (cachedFallbackConfig != null) {
    return cachedFallbackConfig;
}

final Route newFallbackRoute =
        originalRoute.toBuilder()
                     .pathMapping(CatchAllPathMapping.INSTANCE)
                     // Do not propagate an exception raised while resolving a route
                     // to fallback services.
                     .canSetDeferredException(false)
                     .build();
final ServiceConfig newFallbackConfig = fallbackServiceConfig.withRoute(newFallbackRoute);
newFallbackServiceConfigs.put(originalRoute, newFallbackConfig);
return newFallbackConfig;
}
```

Reply:
Fixed. Thanks!!